lib/spack/spack/config.py

```diff
@@ -763,57 +764,93 @@ def _merge_yaml(dest, source):
     with `::` instead of `:`, and the key will override that of the
     parent instead of merging.
 
+    If override is True, then source will be copied to dest unconditionally.
     """
     def they_are(t):
         return isinstance(dest, t) and isinstance(source, t)
 
-    # If both are None, handle specially and return None.
-    if source is None and dest is None:
+    if source is None:
         return None
 
-    # If source is None, overwrite with source.
-    elif source is None:
-        return None
+    if not override:
+        # Source list is prepended (for precedence)
+        if they_are(list):
+            dest[:] = source + [x for x in dest if x not in source]
+            return dest
+
+        # Source dict is merged into dest.
+        elif they_are(dict):
+            # track keys for marking
+            key_marks = {}
+
+            for sk, sv in iteritems(source):
+                if _override(sk) or sk not in dest:
+                    # if sk ended with ::, or if it's new, completely override
+                    dest[sk] = copy.copy(sv)
+                else:
+                    # otherwise, merge the YAML
+                    dest[sk] = _merge_yaml(dest[sk], source[sk], _override(sk))
```

Comment:
`_override(sk)` is guaranteed to be false here (i.e. if it was true we would have executed the "if" branch) and I don't think there's anywhere else that uses the `override` parameter so I think this added parameter can be reverted.